sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala

```diff
@@ -42,15 +42,55 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks {
     InternalRow.fromSeq(values.map(CatalystTypeConverters.convertToCatalyst))
   }
 
+  protected def convertToCatalystUnsafe(a: Any): Any = a match {
+    case arr: Array[Boolean] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Byte] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Short] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Int] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Long] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Float] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Double] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case other => CatalystTypeConverters.convertToCatalyst(other)
+  }
+
   protected def checkEvaluation(
       expression: => Expression, expected: Any, inputRow: InternalRow = EmptyRow): Unit = {
     val serializer = new JavaSerializer(new SparkConf()).newInstance
     val expr: Expression = serializer.deserialize(serializer.serialize(expression))
-    val catalystValue = CatalystTypeConverters.convertToCatalyst(expected)
+    // No codegen version expects GenericArrayData
+    val catalystValue = expected match {
```

Comment:
@cloud-fan I would like to handle this at callee side (i.e. `ExpressionEvalHelper.scala`). If we ask the caller side to pass catalyst array, to change implementation in runtime or code generation requires changes at caller sites. This is because catalyst array (e.g. `GenericArrayData` or `UnsafeArrayData` depend on the implementation of runtime or code generation.

 `expected` is referred to two places. One is for non-codegen result. The other is for codegen result. Since these two parts may require different catalyst arrays (e.g.  `GenericArrayData` or `UnsafeArrayData`), it is not easy for a caller site to pass one catalyst array.

What do you think?